fix: Find peers that can provide state headers

nikurt opened the pull request:
### The problem:

If a node is configured as a chunk-only producer (technically, if it doesn't have `tracked_shards = [0]` in `config.json`), then it can't provide state headers or state parts to other nodes.

### The root cause:

The nodes use archaic logic to set `tracked_shards` in `Handshake`. That logic sets `tracked_shard` to either `[]` or `[0,1,2,3]`.

### Solution:

Ask a random peer for state sync header without taking their `tracked_shards` into account.

nikurt commented on chain/client/src/sync/state.rs:
```diff
@@ -562,6 +562,7 @@ impl StateSync {
         let possible_targets = self.possible_targets(shard_id, highest_height_peers)?;
 
         if possible_targets.is_empty() {
+            tracing::debug!(target: "sync", "Can't request a state header: No possible targets");
```
This can spam the log, but it's important enough.

nikurt commented on chain/client/src/sync/state.rs:
```diff
@@ -602,6 +603,7 @@ impl StateSync {
         new_shard_sync_download: &mut ShardSyncDownload,
     ) {
         let peer_id = possible_targets.choose(&mut thread_rng()).cloned().unwrap();
+        tracing::debug!(target: "sync", ?peer_id, shard_id, ?sync_hash, ?possible_targets, "request_shard_header");
```
This only gets printed every 60 seconds (`state_sync_timeout`)

nikurt commented on chain/epoch-manager/src/shard_tracker.rs:
```diff
@@ -126,11 +126,17 @@ impl ShardTracker {
-            if !is_me {
-                return account_cares_about_shard;
-            } else if account_cares_about_shard {
+            if account_cares_about_shard {
```
Not changing the logic, but make the conditions easier to reason about.

nikurt commented on chain/client/src/client.rs:
```diff
@@ -2718,15 +2718,16 @@ impl Client {
-            (0..num_shards).collect()
-        };
+
+        // TODO: How expensive is it to recompute the set of tracked shards?
```
The archaic logic was here. Replaced with `ShardTracker`.

akhi3030 commented:
Looks like @VanBarbascu, who is a codeowner is reviewing so removing myself.

saketh-are commented:
It looks good to me. However, I have confirmed that `tracked_shards` of a direct peer is shared once upon forming a connection and is not updated again over the lifetime of the connection. Does it mean this PR doesn't really achieve what you hope?

Question: "To provide a state sync header, a node needs to include a chunk from the last epoch of that shard." Does it need to be a chunk for the specific shard the requester is trying to sync? What are the requirements for a node to be able to provide this (is it the same set of nodes which can serve the state parts for that (shard, epoch))?

VanBarbascu reviewed (approved):
Looks good,
As @saketh-are pointed out, this does not solve the staleness of the data but at least it gets the right one.

nikurt commented:
Thanks @saketh-are for confirming that `tracked_shards` is not updated.
Reverted the change to keep the archaic logic.
Changed state sync logic to request state header from random peers.

saketh-are commented on chain/client/src/sync/state.rs:
```diff
@@ -857,32 +837,37 @@ impl StateSync {
-                shard_sync_download.downloads[0].run_me.store(true, Ordering::SeqCst);
-                shard_sync_download.downloads[0].error = false;
-                shard_sync_download.downloads[0].prev_update_time = now;
+            let download_timeout = now - download.prev_update_time > self.timeout;
```
:+1: Nice cleanup, those mutable booleans were ugly

saketh-are commented on chain/client/src/sync/state.rs:
```diff
@@ -857,32 +837,37 @@ impl StateSync {
+                metrics::STATE_SYNC_HEADER_ERROR.with_label_values(&[&shard_id.to_string()]).inc();
+            }
+            // Retry in case of timeout or failure.
+            if download_timeout || download.error {
```
If I've understood it correctly you didn't change the main behavior in this function, you've reorganized it so that you can write metrics.

nikurt replied:
Correct :+1:

nikurt commented:
TODO: Improve this usage of `tracked_shards`: https://github.com/near/nearcore/blob/5ce82b3a20c413346d092cf007f730fef1b731fa/chain/network/src/peer_manager/peer_manager_actor.rs#L823
